apm-agent-plugins/apm-jdk-httpserver-plugin/src/main/java/co/elastic/apm/agent/httpserver/HttpHandlerInstrumentation.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.httpserver;
+
+import co.elastic.apm.agent.sdk.ElasticApmInstrumentation;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import java.util.Collection;
+import java.util.Collections;
+
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+public class HttpHandlerInstrumentation extends ElasticApmInstrumentation {
+
+    @Override
```

Comment:
Here it's relevant to override `getTypeMatcherPreFilter` to ensure that it's only applied within package `com.sun.net.httpserver`, otherwise that makes `hasSuperType` matcher expensive. This will make it work only with the implementations that are part of the JDK.

-----

Alternatively, if you aim to extend support to implementations that might be provided through `com.sun.net.httpserver.spi.HttpServerProvider`, this would require to apply instrumentation at runtime when instances are returned by `HttpServerProvider`. This could easily be done in a follow-up PR unless required for your use-case.

Reply:
Thank you for your review.

I am fine with only instrumenting `com.sun.net.httpserver` `Handler`. But then the `HttpHandlerTest` passes no longer because of `java.lang.NoClassDefFoundError: org/slf4j/LoggerFactory`. The reason for the exception is that `com.sun.net.httpserver.*` is loaded by the platform classloader which cannot load `org/slf4j/LoggerFactory` and neither can the agent class loader, because `org/slf4j` is not listed in `co.elastic.apm.agent.bci.classloading.IndyPluginClassLoaderParent#agentPackages`. This is not an issue in the release build because `org/slf4j`is shaded.

How should I proceed? Remove the logger? Remove the test?